uacp/conn_test.go

```diff
@@ -6,46 +6,59 @@ package uacp
 
 import (
 	"context"
+	"net"
 	"testing"
 	"time"
 
 	"github.com/pascaldekloe/goe/verify"
 )
 
 func TestConn(t *testing.T) {
-	ep := "opc.tcp://127.0.0.1:4840/foo/bar"
-	ln, err := Listen(ep, nil)
-	if err != nil {
-		t.Fatal(err)
-	}
-	defer ln.Close()
-
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-
-	done := make(chan struct{})
-	acceptErr := make(chan error, 1)
-	go func() {
-		c, err := ln.Accept(ctx)
+	t.Run("server exists ", func(t *testing.T) {
+		ep := "opc.tcp://127.0.0.1:4840/foo/bar"
+		ln, err := Listen(ep, nil)
 		if err != nil {
-			acceptErr <- err
-			return
+			t.Fatal(err)
 		}
-		defer c.Close()
-		close(done)
-	}()
+		defer ln.Close()
 
-	if _, err = Dial(ctx, ep); err != nil {
-		t.Error(err)
-	}
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
 
-	select {
-	case <-done:
-	case err := <-acceptErr:
-		t.Fatalf("accept fail: %v", err)
-	case <-time.After(time.Second):
-		t.Fatal("timed out")
-	}
+		done := make(chan struct{})
+		acceptErr := make(chan error, 1)
+		go func() {
+			c, err := ln.Accept(ctx)
+			if err != nil {
+				acceptErr <- err
+				return
+			}
+			defer c.Close()
+			close(done)
+		}()
+
+		if _, err = Dial(ctx, ep); err != nil {
+			t.Error(err)
+		}
+
+		select {
+		case <-done:
+		case err := <-acceptErr:
+			t.Fatalf("accept fail: %v", err)
+		case <-time.After(time.Second):
+			t.Fatal("timed out")
+		}
+	})
+
+	t.Run("Server does not exists", func(t *testing.T) {
+		ep := "opc.tcp://example.com:4840"
```

Comment:
Note that `example.com` resolves to an ip address. While unlikely that they run a server at 4840 it might be nicer to not spam them more. What about `127.0.0.99:56789`?

```sh
$ host example.com
example.com has address 93.184.216.34
example.com has IPv6 address 2606:2800:220:1:248:1893:25c8:1946
example.com mail is handled by 0 .
```

Reply:
The error only occurs when the Address resolves, so for `127.0.0.99:56789` it will return `connect: connection refused`.